Embed recovery. Fixes MER#1987

Thaodan opened the pull request:
Some devices like Sonys X series have no recovery partition on these devices the recovery needs to be inside the regular initramfs.

To do so we check if `skip_initramfs` hasn't been passed (read [line 56](https://github.com/mer-hybris/hybris-initrd/pull/27/files#diff-60e3904a259a7d36d7fb4c5546eb9020ba1eae3598e638333da6083e7f4919aeR56)) and then start the recovery.

Background:
Bootloaders for Android phones pass `skip_init_ramfs` when the system should be booted normally and don't pass it when the key combination for recovery mode was used.
We piggy back on this behavior by using the kernel commandline option is mostly the samen way except that we always use a init-ramfs.

sledges commented on droid-hal-device-img-boot.inc:
```diff
@@ -73,6 +75,12 @@ BuildRequires:  yamui
 BuildRequires:  openssh-server
 
+%if 0%{?embed_recovery:1}
+BuildRequires: shpp
```
Your contribution of shpp is appreciated, but I'm afraid it cannot be justified in the context of HW adaptation recovery, as we are trying to keep dependencies to a minimum. In this particular use-case a `sed` command should be sufficient.

sledges commented on droid-hal-device-img-boot.inc:
```diff
@@ -73,6 +75,12 @@ BuildRequires:  yamui
+%if 0%{?embed_recovery:1}
+BuildRequires: shpp
+%else
+%define embed_recovery 0
```
This renders some parts of the code unreachable, because defining something in .spec (even if equal to 0), counts as defined.
See here: https://backreference.org/2011/09/17/some-tips-on-rpm-conditional-macros/
and here: https://github.com/mer-hybris/droid-hal-device/blob/5fd1782c7f7b898fd66469d6cd43890c82219730/droid-hal-%40DEVICE%40.spec.template#L14

sledges commented on droid-hal-device-img-boot.inc:
```diff
@@ -98,6 +106,7 @@ Requires: oneshot
 %description
 %{summary}
 
+%if 0%{!?embed_recovery:1}
```
As mentioned above, this %if never evaluates to true, because embed_recovery is now always defined (either to 1, or to 0). This also applies to all similar statements against `embed_recovery` below

sledges commented on droid-hal-device-img-boot.inc:
```diff
@@ -169,16 +179,19 @@ sed --in-place 's|@DISPLAY_BRIGHTNESS@|%{display_brightness}|' %{_local_initrd_d
 popd
-%{mkbootimg_cmd} hybris-boot.img
-
+%{mkbootimg_cmd} hybris-boot.img 
```
extra space at the end of the line (makes the diff a bit unclear)

sledges commented on droid-hal-device-img-boot.inc:
```diff
@@ -17,6 +17,8 @@
 # battery_capacity_file:      Path to read current battery charge from.
 # battery_capacity_threshold: Battery threshold for factory reset [0, 100].
 #                             Default 0 (no threshold).
+# embed_recovery: in case the recovery needs to embeded inside
```
nitpicking a typo: "needs to be embedded"

Thaodan commented:
How can I default to an alternative value like in shell Scripts with ${too:-bar} ?

About the dependencies , how would you do this with just sed?
I'm not sure how to do this in a clean way.

sledges commented:
> How can I default to an alternative value like in shell Scripts with ${too:-bar} ?

`%{foo}%{!?foo:bar}`, e.g. `%{embed_recovery}%{!?embed_recovery:0}`

> About the dependencies , how would you do this with just sed? I'm not sure how to do this in a clean way.

`sed "s/@EMBED_RECOVERY@/%{embed_recovery}%{!?embed_recovery:0}/g" -i %{_local_initrd_dir}/jolla-init`

Thaodan commented:
@sledges No problem. I fixed that line that you mean in mksfosinitrd and cleaned that feature a bit up.
About the actual init script to you want the embed recovery code to be present when embed recovery is not used or not present when its not used (shpp solution)?

krnlyng commented on droid-hal-device-img-boot.inc:
```diff
@@ -202,11 +212,13 @@ touch %{buildroot}/lib/modules/%{kernelver}/{modules.order,modules.builtin}
 /boot/hybris-boot.img
 %defattr(644,root,root,-)
-/lib/modules/%{kernelver}
+/Li/modules/%{kernelver}
```
Is this a typo?

Thaodan commented:
Yes fixed.

Olf0 commented:
This was looking well on its way and close to be ready for merging, after the many fixes and improvements above.
@sledges & @Thaodan, is there a specific reason, why this has been stalling for more than 2 years.

Olf0 commented:
The reason why iI am asking, is that
1. The requested changes displayed below are the ones @sledges [requested on 19 Nov 2018](https://github.com/mer-hybris/hybris-initrd/pull/27#pullrequestreview-176357628): AFAICS these have been addressed.  Hence @sledges' open ToDos might be to perform his "I haven't checked the changes in .sh for now" to finish his review and ultimately close the open review process.
2. @Thaodan's [last comment is "Yes fixed." on 22 Jun 2019](https://github.com/mer-hybris/hybris-initrd/pull/27#issuecomment-504692436), hence I assume he sees no open issues with this MR for more than two years now.
3. Because this MR has been stalled for more than two years, meanwhile merges now result in merge conflicts and created the need to re-base.  This will only become worse, if the stalling continues.
4. The Xperia 10 II is the first device officially supported by Jolla, which needs an "embedded recovery".  As its support was announced and deployed in early 2021, resolving this issue (MER#1987) is crucial in order to cease offering the clumsy, current official advice to stubbornly re-flash.

sledges commented:
> This was looking well on its way and close to be ready for merging, after the many fixes and improvements above.
> @sledges & @Thaodan, is there a specific reason, why this has been stalling for more than 2 years?

The reason is that such change needs to be tested on all Jolla's supported devices to avoid regressions, and in case such risk is minimal and/or this functionality is not enabled in a given kernel, the work of updating submodule for each device is still required.

Since we had had a workaround for recovery, this PR was not on a critical path and thus turned into technical debt, due to other priorities.
It has certainly become more actual with releasing Xperia 10 II, and we will adjust prioritisation accordingly as possible. For now please use the workaround [1], which is, agreed, more involving than other Xperias, nevertheless recovery mode is (and should) not be a frequently used feature.

[1] https://jolla.zendesk.com/hc/en-us/articles/360021620740

Olf0 commented:
> > This was looking well on its way and close to be ready for merging, after the many fixes and improvements above.
> > @sledges & @Thaodan, is there a specific reason, why this has been stalling for more than 2 years?
> 
> The reason is that such change needs to be tested on all Jolla's supported devices to avoid regressions, and in case such risk is minimal and/or this functionality is not enabled in a given kernel, the work of updating submodule for each device is still required.
> 
> […] this PR […] turned into technical debt, due to other priorities. It has certainly become more actual with releasing Xperia 10 II, and we will adjust prioritisation accordingly as possible.

… and additionally by releasing SailfishOS for the Xperia 10 III recently.

Thaodan commented:
Summary of what has to be done to use this PR.
1. Rebase the PR (cc me :) )
2. Update the droid-hal-img-boot submodule to use this patch/branch, (`%define embed_recovery`)
3. Make sure that `Ignore skip_initramfs` is still passed but doesn't do anything on your devices kernel. For the Xperia XA2 this PR can be used:
    https://github.com/mer-hybris/android_kernel_sony_msm/pull/26

Both droid-hal and droid-hal-img-boot have to be build with the changes mentioned above, after doing so flashing _hybris_boot.img_ is enough to test if the PR works.

Thaodan commented:
This PR is obsolete for devices using Android 12 and up.
It was never required for devices using a dedicated recovery partition (refer to the graph above for devices starting with Android 13):
https://source.android.com/docs/core/architecture/partitions/generic-boot#figure-2
